OpenTrons/opentrons-api
# feat(protocol-designer): update logic for form error rendering

# Overview

According to designs, some errors that are detected at the form level should be shown at the field
rather than at the top of the form. Here, I provide logic for passing form level errors to their
implicated fields, and update input components to render those errors at the appropriate time for the following step forms:

- heater shaker
- thermocycler (profile and state type)
- pause

TODO: add same logic to other step form types

AUTH-989

## Test Plan and Hands on Testing

- create or open one of the above step forms
- if necessary, toggle a dropdown for input that will be required
- leave or change the input to be empty
- select "save" button
- verify that the form renders errors at the implicated fields rather than the top of the form

## Changelog

- add logic for form error rendering at field rather than top of form
- add utilities for mapping form errors to their fields
- pass down errors in input form field component. Form-level errors take priority over field errors
- add form-level errors for above 3 step form types with required fields
- update tests

## Review requests

- see test plan

## Risk assessment

medium
